File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
    vector<vector<int>>ans;
    vector<int>path;
    vector<int>nums;
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        nums = candidates;
        dfs(0,target);
        return ans;
    }
    void dfs(int start,int remain){
        if(remain==0){ans.push_back(path);return;}
        for(int i=start;i<nums.size();i++){
            int x = nums[i];
            if(x>remain)break;

            path.push_back(x);
            dfs(i,remain-x);
            path.pop_back();
        }
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        path = []
        candidates.sort()
        def bt(star,remain):
            if remain==0:
                res.append(path[:])
                return
            if remain<0:
                return
            for i in range(star,len(candidates)):
                path.append(candidates[i])
                bt(i,remain-candidates[i])
                path.pop()
        bt(0,target)
        return res